### ISSUE TITLE
Implement multi-game tournaments

### ISSUE THREAD

**Issue by irilyth** [buttonmen-dev/buttonmen#462]
It'd be nice to allow players to create and play in multi-game tournaments, in a variety of formats.

**Comment by danlangford** [comment]
Would the team consider allowing another existing, well tested well vetted engine to actually manage the brackets? Challonge.com is free to use with a free api just for allowing game integrations that you speak of. They support single elim, double elim, round robin, Swiss and multi stage tournaments.  I have implemented their API and it works very well. Another site I am using is called Toornament.com as a power user over there (I run lots of events) I have been given access to their API which is in the private testing phase. They support more formats like League play and even 3 stage tournaments. I feel that at the end of the day their service will be better than Challonge. Tournament is also free and API access will be free. Challonge requires no attribution while Toornament only requires a visual "powered by Toornament" attribution. There are many tournament bracket engines out there. These two are the best ones I have found. Most mess up the Swiss format or the tiebreakers and no other sites worth any thing have an API.

I have watched many of these sites start to grow up and grow out and falter on things like correctly implementing Swiss. I would really encourage you guys to leverage the hard work and success of others and not spread your dev team thin trying to learn in intricacies of tiebreaking when somebody else will give you the functionality for free. Both APIs are poised to allow a deep integration such that people may not even know the engine is at another site (besides the Toornament visual attribution) 

However if you do decide to implement these your self and you get stuck on Swiss details or tiebreaking theories I have come to understand these well enough to consult with the team a bit . Good luck 

Also we can get you access to the Toornamnet API if it's not openly available by the time you want to pursue this

**Comment by blackshadowshade** [comment]
I'm not at all averse to leveraging someone else's work. There will obviously need to be some integration work done, but in principle, I'm all for it.

There are other related issues that will need to be assessed simultaneously, like allowing games to be created automatically by a user that is not a player, and flagging games visually as tournament games.

**Comment by irilyth** [comment]
Yeah, my two cents is that, like many complicated-but-cool-sounding things, if you can mock it up and show that it's doable and actually cool, that's great, and if it requires a lot of up-front investment before it's obvious whether it's doable and/or actually cool, that's more risky.

**Comment by irilyth** [comment]
Another tournament style that we should definitely implement: Two stage group + elimination, where the players are first divided into groups, play a round-robin of games in the group, and then the top N winners from each group move on to an elimination bracket. This lets you have elimination-style tournaments that *start* with an arbitrary number of players, and then whittle down to a power-of-two number of players in the group stage, if for whatever reason you want to start a 13-person tournament or whatever. (One where each player plays one of the buttons in a set, for example.)

**Comment by irilyth** [comment]
Things that the old site had, which I don't think are blockers for merging #2131, but IMHO should be high priorities for what next:

- Allowing 32 players.
- Double-elimination tournaments.
- Restrictions on which buttons are available.
- A setting indicating that players will get a random button.

One feature that the old site didn't have, but which I'd really love to see, is a way to indicate that the random buttons should be distributed evenly -- this would be super useful in adoption tournaments, especially if it avoided games where both buttons were the same, in the first round (or that could be an optional setting I suppose).

**Comment by blackshadowshade** [comment]
> Things that the old site had, which I don't think are blockers for merging #2131, but IMHO should be high priorities for what next:
> 
> * Allowing 32 players.

I can add 32 players now, that's trivial.

> * Double-elimination tournaments.

Once we're happy with single elimination tournaments, this will be pretty easy.

> * Restrictions on which buttons are available.

This is a bigger project, but yes, it's on the radar.

> * A setting indicating that players will get a random button.

I think this needs to be part of the button restriction project, since ideally, we'd want to have random selection within a restricted set of buttons.

> One feature that the old site didn't have, but which I'd really love to see, is a way to indicate that the random buttons should be distributed evenly -- this would be super useful in adoption tournaments, especially if it avoided games where both buttons were the same, in the first round (or that could be an optional setting I suppose).

I think we could actually do this by implementing a specific type of tournament---something like BMTournamentSingleEliminationRandomEvenDistribution, subclassed from BMTournamentSingleElimination, where I specify the button selection rules as part of the tournament definition.
